### PR TITLE
Initial implementation of ktx create and extract tools

### PR DESCRIPTION
**Pull request by VaderY**
Briefs:

- Extract: Initial implementation of the ktx extract tool
- Extract: Support for 1D/2D/Cube/3D, Array/Non-Array and Mip/No-Mip textures
- Extract: Support for image selection
- Extract: Support for raw export format
- Extract: Implement transcoding
- Extract: Integrate validation into the extract tool
- Create: Initial implementation of the ktx create tool
- Create: Support for 2D Non-Array No-Mip textures
- Create: Support for raw and limited png,exr import format
- Validate: Implement the remaining checks for typeSize, LevelIndex and SGD
- Validate: Implement loading and transcoding checks
- Validate: Implement glTF KHR_texture_basisu validation mode
- Validate: Implement extra warning if a discrepancy detected between validate and the library
- CTS: Add additional tests for validate, info, extract and create
- CTS: Minor improvements on the runner script
- Tools: Adopt cxxopts for command line argument parsing
- Tools: Implement --help and usage printings
- KTX: Add ZLIB supercompression support
- KTX: Add explicit error flags for KTX_DECOMPRESS_LENGTH_ERROR and KTX_DECOMPRESS_CHECKSUM_ERROR
- KTX: Add numerous checks to detect invalid files
- KTX: Fix library crashes for invalid input files
- KTX: Fix numerous minor issues
- imageio: Cherry-pick imageio
- imageio: Minor fixes, extend format support and cleanup some warnings
- cxxopts: As this is a single header library we only included the header (and cmake files) instead of a subrepo

What will come later:

- Extract: png/exr support and certain command line argument validation checks
- Create: The missing features and cleanup. Any commented code will be implemented and cleaned up
- For usages/helps there will be additional documentation and descriptions. The current pass focused on the functionality
- Invalid input file testing for info tool

For extracting multiple images at once with '--all' or any 'all' arg the following naming format was chosen:
	`provided_output_path/output_level{}_face{}_layer{}_depth{}.extension` where:
	`_level{}` part is only present if the source texture has more than 1 level
	`_face{}` part is only present if the source texture is cubemap or cubemap array
	`_layer{}` part is only present if the source texture is an array texture
	`_depth{}` part is only present if the source texture baseDepth is more than 1
	where `{}` is replaced with the numeric index of the given component starting from 0
	Note: The inclusion of the optional parts are determined by the source texture regardless of which images are requested

Please review these changes.

### CODE REVIEW

**Comment by MarkCallow** [comment]
Is there any way to get a diff of the "minor fixes" you made to imageio? As the imageio lib is new to this branch in this PR, all its files are shown in their entirety in the PR diff.

**Comment by VaderY** [comment]
> Is there any way to get a diff of the "minor fixes" you made to imageio? As the imageio lib is new to this branch in this PR, all its files are shown in their entirety in the PR diff.

Yes, I kept the changes in 3 separate commit so these changes can be tracked. You can check tools/imageio folder's diff only in the modification commit to see only the changes without the cherry-pick itself: [e15972b5385add105026f8487e6dcaef3be7c1af](https://github.com/KhronosGroup/KTX-Software/pull/677/commits/e15972b5385add105026f8487e6dcaef3be7c1af).

**Comment by MarkCallow** [comment]
> It looks like a Git/CI environment related issue and every other job was successfully on that step.

I'm re-running those 2 jobs. That issue started happening recently and I have not yet had enough energy to engage with Travis support over it. The builds usually work when re-run.

**Comment by VaderY** [comment]
> I'm re-running those 2 jobs.

Thanks, looks like it worked.

I know you will be away for a couple of days, but once you have the time, please review this PR as the next one is around the corner too. I would recommend to dont spend too much time on this version of the 'create' tool as the next PR will revamp it.

**Review by MarkCallow — CHANGES_REQUESTED** [review]
I've reviewed everything but the create command. I've requested a few minor fixes.

I notice that `extract` always processes all transcode options. When you add support for png output, it should bail immediately if the transcode option is specified unless the target is one of the uncompressed formats. This set me thinking that perhaps KTX2 should be one of the supported output formats so you can extract, e.g. a level, transcode it to a block compressed format and save it as a usable file.

What was the fix that necessitated updating the various files in `tests/testimages`?

File: tools/ktx/command.cpp [hunk 1]
```diff
@@ -5,208 +5,116 @@
...
+std::string Command::version() const {
+    return testrun ? STR(KTX_DEFAULT_VERSION) : STR(KTX_VERSION);
 }
 
-void Command::writeId(std::ostream& dst, bool chktest) {
-    dst << processName << " ";
-    dst << (!chktest || !genericOptions.test ? version : defaultVersion);
+// -------------------------------------------------------------------------------------------------
+
+void CommandWithFormat::initOptions(cxxopts::Options& opts) {
```
[the remark this answers is not in the document]
Comment (VaderY) [reply tools/ktx/command.cpp:91]: You are right, I have already improved the naming and structure of the options, Next PR will contain the fix

File: tools/imageio/formatdesc.h [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+        }
+        if (m == KHR_DF_MODEL_YUVSDA && channelCount == 2) {
+            samples[1].channelType = KHR_DF_CHANNEL_YUVSDA_ALPHA;
+        }
+    }
+
+    /// @brief Constructor for non-compressed textures with varying bit lengths
+    ///        or channel types and custom sampleLower and sampleUpper.
+    ///
+    /// Each channle has the same basic data type.  Use this for unnormalized
```
Comment (MarkCallow) [inline tools/imageio/formatdesc.h:387]: Typo. s/channle/channel/

**Comment by VaderY** [comment]
I pushed the resolutions, but as I can see the mentioned CI issues are coming up

File: lib/strings.c [hunk 3]
```diff
@@ -35,9 +35,11 @@ static const char* const errorStrings[] = {
     "Out of memory.",                                 /* KTX_OUT_OF_MEMORY */
     "Transcoding of block compressed texture failed.",/* KTX_TRANSCODE_FAILED */
     "Not a KTX file.",                                /* KTX_UNKNOWN_FILE_FORMAT */
-    "Texture type not supported.",      /* KTX_UNSUPPORTED_TEXTURE_TYPE */
+    "Texture type not supported.",                    /* KTX_UNSUPPORTED_TEXTURE_TYPE */
```
Comment (MarkCallow) [inline lib/strings.c:38]: Where is KTX_UNSUPPORTED_TEXTURE_TYPE still used?

Comment (VaderY) [reply lib/strings.c:38]: It has around a dozen uses like:
On ktxFormatSize_initFromDfd failure
In _ktxUnpackETC if the format is not supported
In ktxTexture_GLUploadPrivate

**Review by MarkCallow — APPROVED** [review]

**Pull request merged**